File: pattern/csharp/demo_delegate/demo_delegate/service/ProductService.cs
using System;
using demo_delegate.model;
using System.Collections.Generic;
using System.Collections;

namespace demo_delegate.service
{
    public class ProductService
    {
        private List<Product> products = new List<Product>();

        public ProductService()
        {
            products.Add(new Product(1, "iphone", "gold", 15, 1000));
            products.Add(new Product(2, "nokia", "yellow", 10, 500));
            products.Add(new Product(3, "samsung", "gold", 8, 700));
            products.Add(new Product(4, "oppo", "red", 5, 300));
            products.Add(new Product(5, "lg", "blue", 120, 900));
        }

        public List<Product> getAll()
        {
            return products;
        }

        public List<Product> search(search search)
        {
            return search.Invoke();
        }
    }
}

File: pattern/java/src/service/ProductService.java
package service;

import model.Product;

import java.util.ArrayList;
import java.util.List;

public class ProductService {
    private static List<Product> products= new ArrayList<>();

    static {
        products.add(new Product(1,"iphone", "gold", 15, 1000));
        products.add(new Product(2,"nokia", "yellow", 10, 500));
        products.add(new Product(3,"samsung", "gold", 8, 700));
        products.add(new Product(4,"oppo", "red", 5, 300));
        products.add(new Product(5,"lg", "blue", 120, 900));
    }


    public List<Product> search(ProductSearch productSearch){
        return productSearch.search();
    }

    public List<Product> getAll(){
        return products;
    }
}
